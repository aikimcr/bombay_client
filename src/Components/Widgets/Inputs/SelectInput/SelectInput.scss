.select-input {
  font-size: var(--standard-font-size);

  &__select {
    width: 100%;
    font-size: var(--standard-font-size);
    padding: 1px .75em;
    cursor: pointer;
  }
}