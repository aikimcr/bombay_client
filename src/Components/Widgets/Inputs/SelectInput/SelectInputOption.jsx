import React, { useRef } from 'react';
import PropTypes from 'prop-types';

export const SelectionInputOption = (props) => {
  return <option value={props.value}>{props.label}</option>;
};

SelectionInputOption.propTypes = {
  value: PropTypes.any.isRequired,
  label: PropTypes.string,
};
