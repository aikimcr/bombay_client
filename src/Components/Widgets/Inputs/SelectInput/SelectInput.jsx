import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { SelectionInputOption } from './SelectInputOption';

import './SelectInput.scss';

export const SelectInput = (props) => {
  const topRef = useRef(null);

  const modelId = props.model ? props.model.get('id') : 'isNew';
  const unsetLabel = props.unsetLabel || `Please Select ${props.labelText}`;
  const modelValue = props.model ? props.model.get(props.fieldName) : '';
  const defaultValue = modelValue == null ? '' : modelValue;

  return (
    <div ref={topRef} className="select-input" data-testid="select-input">
      <select
        id={props.id}
        className="select-input__select"
        data-testid="select-input-select"
        name={props.name}
        defaultValue={defaultValue}
        onChange={props.onChange}
      >
        <SelectionInputOption value="" label={unsetLabel} />
        {props.options.map((option) => {
          const key = `${props.id}-${option.value}`;
          const label = option.label || option.value;

          return (
            <SelectionInputOption
              key={key}
              value={option.value}
              label={label}
            />
          );
        })}
      </select>
    </div>
  );
};

SelectInput.propTypes = {
  name: PropTypes.string,
  options: PropTypes.array.isRequired,
  unsetLabel: PropTypes.string,
  onChange: PropTypes.func,
};
