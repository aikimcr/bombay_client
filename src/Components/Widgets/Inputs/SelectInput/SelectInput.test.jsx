import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import { SelectInput } from './SelectInput';

jest.useFakeTimers();

const options = [
  {
    value: 'bird',
  },
  {
    value: 'plane',
  },
  {
    value: 'jet',
  },
  {
    value: 'rocket',
  },
  {
    value: 'bat',
  },
];

it('Component should match snapshot', () => {
  const { asFragment } = render(
    <SelectInput
      modelName="xyzzy"
      fieldName="plover"
      labelText="plugh"
      options={options}
    />,
  );

  expect(asFragment).toMatchSnapshot();
});

it('should set the default value to the field value', () => {
  const model = {
    get: jest.fn((fieldName) => {
      return 'bird';
    }),
  };

  const result = render(
    <SelectInput
      name="plover"
      options={options}
      model={model}
      defaultValue="bird"
    />,
  );

  const selectInput = screen.getByTestId('select-input-select');
  expect(selectInput.value).toEqual('bird');
});

it('should call onChange', () => {
  const changeHandler = jest.fn();

  const result = render(
    <SelectInput
      name="plover"
      options={options}
      onChange={changeHandler}
      defaultValue="jet"
    />,
  );

  const selectInput = screen.getByTestId('select-input-select');

  act(() => {
    fireEvent.change(selectInput, { target: { value: 'bat' } });
  });

  expect(changeHandler.mock.calls.length).toBe(1);
  expect(changeHandler.mock.calls[0].length).toBe(1);
});
