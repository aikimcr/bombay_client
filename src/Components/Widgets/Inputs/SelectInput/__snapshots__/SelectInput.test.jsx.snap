// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `[Function]`;
